### DISCUSSION TITLE
How to transform the validation dataset?

### DISCUSSION THREAD

**Discussion by M-Amrollahi** [ultralytics/ultralytics#4646] [Q&A] [answered]
I have modified the ```Albumentation``` class from the source code (/ultralytics/data/augment.py) to have my own data augmentation.

How can I make appropriate data pipeline for validation dataset?

**Comment by HtwoOtwo** [comment]
I think you can try to modify the mode when getting dataloader (mode='val' --> mode='train' ):
`self.test_loader = self.get_dataloader(self.testset, batch_size=batch_size * 2, rank=-1, mode='train')`
'mode' decides whether to enable augmentation.

if you want to customize your own data augmentation pipeline, mybe you can refer to `build_transforms` function [https://github.com/ultralytics/ultralytics/blob/d2cf7acce0dde0a5aece4b4e50132c40a09acd23/ultralytics/data/dataset.py#L135C9-L135C25](url)

**Reply by M-Amrollahi** [reply]
Actually, the link does not work.
I can transform the images before I feed them to model during validation, but I do not know whether any transformation will apply to images inside prediction operation.

**Reply by HtwoOtwo** [reply]
the link: https://github.com/ultralytics/ultralytics/blob/d2cf7acce0dde0a5aece4b4e50132c40a09acd23/ultralytics/data/dataset.py#L135
I don't know how you apply the transform to your images, but as far as I know, the transform should be completed when the dataset is created. These transform will be applied when `__getitem__` works https://github.com/ultralytics/ultralytics/blob/d2cf7acce0dde0a5aece4b4e50132c40a09acd23/ultralytics/data/base.py#L238

**Reply by HtwoOtwo** [reply]
#4658 this issue may also help you.

**Reply by HtwoOtwo** [reply]
I think the easiest way is to replace dataset.transforms to your own transforms when you create an object of validation dataset. e.g.
```
self.test_loader = self.get_dataloader(self.testset, batch_size=batch_size * 2, rank=-1, mode='train')
self.test_loader.dataset.transforms = your transformers generated by build_transforms() # see above link
```

**Reply by pderrenger** [reply]
@HtwoOtwo to apply custom transformations to your validation dataset in YOLOv8, you can modify the dataset's `transforms` attribute with your own transformation pipeline. Here's a concise example of how you might do this:

```python
from ultralytics import YOLO, build_transforms

# Load your model
model = YOLO('yolov8n.pt')

# Get your validation dataloader
val_loader = model.val_dataloader

# Define your custom transformations using build_transforms
custom_transforms = build_transforms(augment=True, visual=True)  # Set your desired augmentations

# Apply your custom transformations to the validation dataset
val_loader.dataset.transforms = custom_transforms
```

Make sure to define your custom transformations using the `build_transforms` function or any other method you prefer, ensuring it is compatible with the dataset's expected input format. The `augment` and `visual` parameters are placeholders and should be set according to your specific augmentation needs.

Remember that the validation phase typically does not include the same level of data augmentation as the training phase, as it aims to evaluate the model's performance on data that is representative of real-world conditions. However, if you have a specific reason to apply certain transformations during validation, the above method allows you to customize the pipeline accordingly.

**Comment by M-Amrollahi** [comment]
@glenn-jocher 
1- Which data normalization has been applied to initial Yolov8 when it needed to be trained for first time? Like:
```
T.Normalize(...)
or
img = img / 255.0
```

2- Which transformation will be applied to an image during prediction? As I normalized my images on my own way with albumentation, I need to know where should I apply transformation for val set.

**Reply by glenn-jocher** [reply, marked as answer]
@M-Amrollahi hello! For YOLOv8, the data normalization and transformations applied during training and validation are defined in the dataset configuration YAML file and the augmentation code. By default, images are scaled to the range [0, 1] by dividing by 255.0, and various augmentations may be applied during training.

1. The default normalization for YOLOv8 is typically `img / 255.0`, scaling pixel values to [0, 1].

2. During prediction (inference), the same normalization (`img / 255.0`) is applied, and no additional augmentations are usually performed. However, the image is resized to the model's input size while maintaining the aspect ratio, and padding is added if necessary.

If you've customized the `Albumentation` class for your own data augmentation, you can apply similar transformations to your validation set by modifying the validation pipeline in the dataset configuration YAML file or directly in the code where the validation dataset is loaded and processed. Ensure that the validation transformations are consistent with the model's expected input.

For the validation set, it's important to apply the same normalization as during training but typically without the random augmentations (like flips, rotations, etc.) to maintain the integrity of the validation process.

If you need to apply custom transformations to the validation set, you can do so in the `load_mosaic()` and `load_image()` functions in the `datasets.py` file, where the validation images are loaded. Make sure to apply the transformations before the images are passed to the model for validation.

Remember to keep the validation transformations consistent with the training pipeline for accurate performance evaluation. If you have further questions or need assistance with specific code changes, feel free to ask!
